Backport of upstream strukturag/libheif PR #1680 "Support svt-av1 4.0.0"
(merge commit 2a7a383ffe90a5d36d7e0c939e6a7ae953e6ba55, merged 2026-04-08).
Fixes FTBFS against SVT-AV1 4.x: the 1.21.2 sequence-encoding path references
the SvtAv1PredStructure enum unguarded, which SVT-AV1 4.x no longer declares
(pred_structure is a plain uint8_t there). Adds the missing
SVT_AV1_CHECK_VERSION guards and enables upstream's intended AVIF
still-image mode on 4.0+. Same fix BLFS applies to libheif 1.21.2
(BLFS ticket #22839). Applies clean (zero fuzz) to the pinned 1.21.2 tarball.

diff --git a/libheif/plugins/encoder_svt.cc b/libheif/plugins/encoder_svt.cc
index 03d87ca1ae..3e614908fe 100644
--- a/libheif/plugins/encoder_svt.cc
+++ b/libheif/plugins/encoder_svt.cc
@@ -838,8 +838,32 @@ static heif_error svt_start_sequence_encoding_intern(void* encoder_raw, const he
     // svt_config.force_key_frames = true; TODO: this does not seem to work (see all [1])
 
 #if 1
+#if SVT_AV1_CHECK_VERSION(4, 1, 0)
+    switch (options->gop_structure) {
+      case heif_sequence_gop_structure_intra_only:
+        //svt_config.pred_structure = 1; // LOW_DELAY
+        break;
+      case heif_sequence_gop_structure_lowdelay:
+        //svt_config.pred_structure = 1; // LOW_DELAY
+        break;
+      case heif_sequence_gop_structure_unrestricted:
+        svt_config.pred_structure = PredStructure::RANDOM_ACCESS;
+        break;
+    }
+#elif SVT_AV1_CHECK_VERSION(4, 0, 0)
+    switch (options->gop_structure) {
+      case heif_sequence_gop_structure_intra_only:
+        //svt_config.pred_structure = 1; // LOW_DELAY
+        break;
+      case heif_sequence_gop_structure_lowdelay:
+        //svt_config.pred_structure = 1; // LOW_DELAY
+        break;
+      case heif_sequence_gop_structure_unrestricted:
+        svt_config.pred_structure = 2; // RANDOM_ACCESS
+        break;
+    }
+#else
     // TODO: setting pref_structure to SVT_AV1_PRED_LOW_DELAY_B hangs the encoder
-
     switch (options->gop_structure) {
       case heif_sequence_gop_structure_intra_only:
         //svt_config.pred_structure = SvtAv1PredStructure::SVT_AV1_PRED_LOW_DELAY_B;
@@ -851,6 +875,7 @@ static heif_error svt_start_sequence_encoding_intern(void* encoder_raw, const he
         svt_config.pred_structure = SvtAv1PredStructure::SVT_AV1_PRED_RANDOM_ACCESS;
         break;
     }
+#endif
 
     if (options->keyframe_distance_max) {
       svt_config.intra_period_length = options->keyframe_distance_max; // TODO -1 ?
@@ -858,9 +883,10 @@ static heif_error svt_start_sequence_encoding_intern(void* encoder_raw, const he
 #endif
   }
   else {
-    // TODO: enable when https://gitlab.com/AOMediaCodec/SVT-AV1/-/issues/2245 is resolved
-    // svt_config.avif = true;
-    // encoder->still_image_mode = true;
+#if SVT_AV1_CHECK_VERSION(4, 0, 0)
+    svt_config.avif = true;
+    encoder->still_image_mode = true;
+#endif
   }
 
   if (color_format == EB_YUV422 || bitdepth_y > 10) {
@@ -920,7 +946,7 @@ static heif_error read_encoder_output_packets(void* encoder_raw, bool done_sendi
 
         /* TODO: this is a hack
          * When using
-         *           svt_config.pred_structure = SvtAv1PredStructure::SVT_AV1_PRED_LOW_DELAY_B;
+         *           svt_config.pred_structure = 1; // LOW_DELAY
          * svt_av1_enc_get_packet() hangs on the second call. As a workaround, we can leave the
          * loop when we got the image:
          */
